File: leetcode/C++/1026_maximum_difference_between_node_and_ancestor.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
    int helper(TreeNode * root, int curmax, int curmin) {
        if (root == nullptr) return curmax - curmin;
        curmax = max(root->val, curmax);
        curmin = min(root->val, curmin);
        int l = helper(root->left, curmax, curmin);
        int r = helper(root->right, curmax, curmin);
        return max(l, r);
    }
public:
    int maxAncestorDiff(TreeNode* root) {
        if (root == nullptr) return 0;
        return helper(root, root->val, root->val);
    }
};

File: leetcode/Golang/1026_maximum_difference_between_node_and_ancestor.go
/*
* @Author: Teddyky
* @Date:   2020-11-10 13:32:32
* @Last Modified by:   Teddyky
* @Last Modified time: 2020-11-10 13:32:42
*/
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxAncestorDiff(root *TreeNode) (ans int) {
    helper(root, root.Val, &ans)
    // temp := max(abs(root.Val - minv), abs(root.Val - maxv))
    // ans = max(ans, temp)
    return
}

func min(a, b int) int {
    if a < b { return a }
    return b
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

func abs(num int) int {
    if num < 0 { return -num }
    return num
}

func helper(root *TreeNode, val int, ans *int) (int, int) {
    if root != nil {
        leftMin, leftMax := helper(root.Left, root.Val, ans)
        rightMin, rightMax := helper(root.Right, root.Val, ans)
        minv, maxv := min(leftMin, rightMin), max(leftMax, rightMax)
        minv, maxv = min(minv, val), max(maxv, val)
        temp := max(abs(val - minv), abs(val - maxv))
        *ans = max(*ans, temp)
        return minv, maxv
    }
    return val, val
} 

